service/frontend/http_api_server.go

```diff
@@ -0,0 +1,438 @@
+		_, _ = w.Write([]byte(`{"code": 13, "message": "failed to marshal error message"}`))
+		return
+	}
+
+	w.WriteHeader(runtime.HTTPStatusFromCode(s.Code()))
+	_, _ = w.Write(buf)
+}
+
+func (h *HTTPAPIServer) newMarshaler(pretty bool, disablePayloadShorthand bool) runtime.Marshaler {
+	marshalOpts := proxy.JSONPBMarshalerOptions{DisablePayloadShorthand: disablePayloadShorthand}
+	if pretty {
+		marshalOpts.Indent = "  "
+	}
+	unmarshalOpts := proxy.JSONPBUnmarshalerOptions{DisablePayloadShorthand: disablePayloadShorthand}
+	if m, err := proxy.NewJSONPBMarshaler(marshalOpts); err != nil {
+		panic(err)
+	} else if u, err := proxy.NewJSONPBUnmarshaler(unmarshalOpts); err != nil {
+		panic(err)
+	} else {
+		return proxy.NewGRPCGatewayJSONPBMarshaler(m, u)
+	}
+}
+
+func (h *HTTPAPIServer) incomingHeaderMatcher(headerName string) (string, bool) {
+	// Try ours before falling back to default
+	if h.matchAdditionalHeaders[headerName] {
+		return headerName, true
+	}
+	return runtime.DefaultHeaderMatcher(headerName)
+}
+
+// inlineClientConn is a [grpc.ClientConnInterface] implementation that forwards
+// requests directly to gRPC via interceptors. This implementation moves all
+// outgoing metadata to incoming and takes resulting outgoing metadata and sets
+// as header. But which headers to use and TLS peer context and such are
+// expected to be handled by the caller.
+type inlineClientConn struct {
+	methods     map[string]*serviceMethod
+	interceptor grpc.UnaryServerInterceptor
+}
+
+var _ grpc.ClientConnInterface = (*inlineClientConn)(nil)
+
+type serviceMethod struct {
+	info    grpc.UnaryServerInfo
+	handler grpc.UnaryHandler
+}
+
+var contextType = reflect.TypeOf((*context.Context)(nil)).Elem()
+var protoMessageType = reflect.TypeOf((*proto.Message)(nil)).Elem()
+var errorType = reflect.TypeOf((*error)(nil)).Elem()
+
+func newInlineClientConn(servers map[string]any, interceptors []grpc.UnaryServerInterceptor) *inlineClientConn {
+	// Create the set of methods via reflection. We currently accept the overhead
+	// of reflection compared to having to custom generate gateway code.
+	methods := map[string]*serviceMethod{}
+	for qualifiedServerName, server := range servers {
+		serverVal := reflect.ValueOf(server)
+		for i := 0; i < serverVal.NumMethod(); i++ {
+			methodVal := serverVal.Method(i)
+			// We assume the methods we want only accept a context + request and only
+			// return a response + error. We also assume the method name matches the
+			// RPC name.
+			methodType := methodVal.Type()
+			validRPCMethod := methodType.Kind() == reflect.Func &&
+				methodType.NumIn() == 2 &&
+				methodType.NumOut() == 2 &&
+				methodType.In(0) == contextType &&
+				methodType.In(1).Implements(protoMessageType) &&
+				methodType.Out(0).Implements(protoMessageType) &&
+				methodType.Out(1) == errorType
+			if !validRPCMethod {
+				continue
+			}
+			fullMethod := "/" + qualifiedServerName + "/" + serverVal.Type().Method(i).Name
```

Comment:
these indexes are guaranteed to line up? it looks like yes though the reflect docs are not 100% clear

you could loop over the type methods and look up the value method by name?

Reply:
They are because `value.NumMethod()` is basically `value.Type().NumMethod()`, see https://github.com/golang/go/blob/449ef3795d8054faf4a601d8d1aab1f624b822f5/src/reflect/value.go#L2037. But yes, no guarantee. I can change.

Reply:
Fixed